File: 0212-word-search-ii/0212-word-search-ii.cpp
class Trie {
private:
    unordered_map<char, Trie*> children;
    bool isEnd;
    string word;
public:
    Trie() {
        isEnd = false;
    }
    Trie* GetChild(char item) {
        return this->children[item];
    }
    void NewChild(char item) {
        this->children[item] = new Trie();
    }
    bool HasChild(char item) {
        if (this->children.count(item)) {
            return true;
        }
        return false;
    }
    bool GetEnd() {
        return this->isEnd;
    }
    void SetEnd(bool value) {
        this->isEnd = value;
    }
    void SetWord(string value) {
        this->word = value;
    }
    string GetWord() {
        return this->word;
    }
};

class Solution {
private:
    Trie* root = new Trie();
public:
    void AddWord(std::string word) {
        Trie* current = root;
        for (auto w : word) {
            if (!current->HasChild(w)) {
                current->NewChild(w);
            }
            current = current->GetChild(w);
        }
        current->SetEnd(true);
        current->SetWord(word);
    }
    void Dfs(int r, int c, int rows, int cols, Trie* root, vector<vector<char>>& board, vector<string>& answer) {
        if (r < 0 || r >= rows || c < 0 || c >= cols || board[r][c] == '#' || !root->HasChild(board[r][c])) {
            return;
        }
        char item = board[r][c];
        board[r][c] = '#';
        root = root->GetChild(item);
        if (root->GetEnd()) {
            answer.push_back(root->GetWord());
            root->SetEnd(false);
        }
        Dfs(r - 1, c, rows, cols, root, board, answer);
        Dfs(r + 1, c, rows, cols, root, board, answer);
        Dfs(r, c - 1, rows, cols, root, board, answer);
        Dfs(r, c + 1, rows, cols, root, board, answer);
        board[r][c] = item;
    }
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        for (auto w : words) {
            AddWord(w);
        }
        int rows = board.size();
        int cols = board[0].size();
        vector<string> answer;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                Dfs(i, j, rows, cols, root, board, answer);
            }
        }
        return answer;
    }
};

File: 0212-word-search-ii/0212-word-search-ii.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end = False
    
    def add_word(self, word):
        root = self
        for c in word:
            if c not in root.children:
                root.children[c] = TrieNode()
            root = root.children[c]
        root.end = True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        for w in words:
            root.add_word(w)
        rows, cols = len(board), len(board[0])
        ans = []
        word = []
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        
        def dfs(r, c, root):
            if not (0 <= r < rows) or not (0 <= c < cols) or board[r][c] == "#":
                return
            char = board[r][c]
            if char not in root.children:
                return
            board[r][c] = "#"
            root = root.children[char]
            word.append(char)
            if root.end:
                ans.append("".join(word))
                root.end = False
            for x, y in directions:
                dfs(r + x, c + y, root)
            word.pop()
            board[r][c] = char
        
        for r in range(rows):
            for c in range(cols):
                dfs(r, c, root)
        
        return ans